tests/doublylinkedlist_ut/doublylinkedlist_ut.c

```diff
@@ -414,5 +465,122 @@ TEST_FUNCTION_CLEANUP(TestMethodCleanup)
 
     }
 
+    /*Tests_SRS_DLIST_43_001: [If listHead is NULL, DList_ForEach shall fail and return a non - zero value.]*/
+    /*Tests_SRS_DLIST_43_012 : [If there are any failures, DList_ForEach shall fail and return a non - zero value.]*/
+    TEST_FUNCTION(DList_ForEach_fails_with_NULL_listHead)
+    {
+        ///arrange
+
+        ///act
+        int result = DList_ForEach(NULL, test_action_function, test_action_context);
+
+        ///assert
+        ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+        ASSERT_ARE_NOT_EQUAL(int, 0, result);
+    }
+    
+    /*Tests_SRS_DLIST_43_002 : [If actionFunction is NULL, DList_ForEach shall fail and return a non - zero value.]*/
+    /*Tests_SRS_DLIST_43_012 : [If there are any failures, DList_ForEach shall fail and return a non - zero value.]*/
+    TEST_FUNCTION(DList_ForEach_fails_with_NULL_actionFunction)
+    {
+        ///arrange
+
+        ///act
+        int result = DList_ForEach(test_pdlist_entry, NULL, test_action_context);
+
+        ///assert
+        ASSERT_ARE_EQUAL(char_ptr, umock_c_get_expected_calls(), umock_c_get_actual_calls());
+        ASSERT_ARE_NOT_EQUAL(int, 0, result);
+    }
+
+    /*Tests_SRS_DLIST_43_009 : [DList_ForEach shall call actionFunction on each entry in the list defined by listHead along with actionContext.]*/
+    /*Tests_SRS_DLIST_43_011 : [DList_ForEach shall succeed and return zero.]*/
+    TEST_FUNCTION(DList_ForEach_succeeds)
+    {
+        ///arrange
+        DLIST_ENTRY head = { 0 };
+        DList_InitializeListHead(&head);
+        DList_InsertTailList(&head, &(simp1.link));
```

Comment:
>Tail [](http://example.com/codeflow?start=20&length=4)

I've always found funny how the fields that hold the dlist are called flink/blink, but we use Tail And Head.

InsetAtFlink :)

./disconsider this comment of mine.

Reply:
Head or tail is for coin tosses